### PR TITLE
update aks cli extension to include run-command feature

### PR DESCRIPTION
**Pull request by haitch**
---

Add run-command sub-command to aks-preview extension.

- az aks **command invoke**
- az aks **command result**

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [x] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
- [x] Have you run `python scripts/ci/test_index.py -q` locally?

For new extensions:

- [ ] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/extensions/extension_summary_guidelines.md).

### About Extension Publish

There is a pipeline to automatically build, upload and publish extension wheels.  
Once your PR is merged into master branch, a new PR will be created to update `src/index.json` automatically.  
The precondition is to put your code inside this repo and upgrade the version in the PR but do not modify `src/index.json`.

### CODE REVIEW

File: src/aks-preview/azext_aks_preview/_help.py [hunk 1]
```diff
@@ -573,6 +573,18 @@
     short-summary: Display diagnostic results for the Kubernetes cluster after kollect is done.
 """
 
+helps['aks command'] = """
```
[the remark this answers is not in the document]
Comment (haitch) [reply src/aks-preview/azext_aks_preview/_help.py]: talked with PM, decide to use **command invoke** and **command result**

File: src/aks-preview/azext_aks_preview/custom.py [hunk 2]
```diff
@@ -1940,6 +1943,32 @@ def aks_upgrade(cmd,    # pylint: disable=unused-argument, too-many-return-state
 
     return sdk_no_wait(no_wait, client.create_or_update, resource_group_name, name, instance)
 
+def aks_runcommand(cmd, client, resource_group_name, name, command_string="", command_files=None):
+    colorama.init()
+
+    if len(command_string) == 0:
```
[the remark this answers is not in the document]
Comment (haitch) [reply src/aks-preview/azext_aks_preview/custom.py]: no to first question

[the remark this answers is not in the document]
Comment (haitch) [reply src/aks-preview/azext_aks_preview/custom.py]: will add unittest

File: src/aks-preview/azext_aks_preview/custom.py [hunk 3]
```diff
@@ -2059,6 +2063,94 @@ def aks_upgrade(cmd,    # pylint: disable=unused-argument, too-many-return-state
...
+    mc = client.get(resource_group_name, name)
+
+    if not command_string:
+        raise CLIError('Command cannot be empty.')
+
+    request_payload = RunCommandRequest()
+    request_payload.command = command_string
+    request_payload.context = _get_command_context(command_files)
+    if mc.aad_profile is not None and mc.aad_profile.managed:
+        request_payload.cluster_token = _get_dataplane_aad_token(cmd.cli_ctx, "6dae42f8-4368-4678-94ff-3960e28e3630")
```
Comment (zqingqing1) [inline src/aks-preview/azext_aks_preview/custom.py:2078]: what is this? can we define a const?

File: src/aks-preview/azext_aks_preview/custom.py [hunk 4]
```diff
@@ -2059,6 +2063,94 @@ def aks_upgrade(cmd,    # pylint: disable=unused-argument, too-many-return-state
...
+def _get_dataplane_aad_token(cli_ctx, serverAppId):
+    # this function is mostly copied from keyvault cli
+    import adal
+    try:
+        return Profile(cli_ctx=cli_ctx).get_raw_token(serverAppId)[0][2].get('accessToken')
+    except adal.AdalError as err:
+        # pylint: disable=no-member
+        if (hasattr(err, 'error_response') and
+                ('error_description' in err.error_response) and
+                ('AADSTS70008:' in err.error_response['error_description'])):
```
Comment (zqingqing1) [inline src/aks-preview/azext_aks_preview/custom.py:2148]: define a const as well.

File: src/aks-preview/azext_aks_preview/custom.py [hunk 5]
```diff
@@ -2059,6 +2063,94 @@ def aks_upgrade(cmd,    # pylint: disable=unused-argument, too-many-return-state
...
+    zipFile.close()
+
+    return str(base64.encodebytes(zipStream.getbuffer()), "utf-8")
+
+
+def _get_dataplane_aad_token(cli_ctx, serverAppId):
+    # this function is mostly copied from keyvault cli
+    import adal
+    try:
+        return Profile(cli_ctx=cli_ctx).get_raw_token(serverAppId)[0][2].get('accessToken')
```
Comment (zqingqing1) [inline src/aks-preview/azext_aks_preview/custom.py:2143]: [0][2]? and we make sure it won't out of index?

**Review by zqingqing1 — CHANGES_REQUESTED** [review]
1. please add tests for your new commands.
2. remember to bump version at setup.py and add release note at history.MD

File: src/aks-preview/azext_aks_preview/custom.py (hunk 3 above)
Comment (haitch) [reply src/aks-preview/azext_aks_preview/custom.py:2078]: this is AKS AAD ServerApp ID, for all AAD enabled aks cluster.

Comment (haitch) [reply src/aks-preview/azext_aks_preview/custom.py:2078]: sure, I can defined this as const.

File: src/aks-preview/azext_aks_preview/custom.py (hunk 5 above)
Comment (haitch) [reply src/aks-preview/azext_aks_preview/custom.py:2143]: 1. this is copied from keyvault CLI, I assume this is well tested.
2. it does have a try block covers it.

**Review by zqingqing1 — APPROVED** [review]
lgtm

**Review by JunSun17 — APPROVED** [review]

**Pull request merged**